ReactWindows/ReactNative/UIManager/UIViewOperationQueueCollection.cs

```diff
@@ -0,0 +1,542 @@
+                {
+                    queueInstance = new UIViewOperationQueueInstance(
+                        _reactContext, new NativeViewHierarchyManager(_viewManagerRegistry, rootViewDispatcher), reactChoreographer),
+                    rootViewCount = 1
+                };
+
+                // Add new tuple to map
+                _dispatcherToOperationQueueInfo.Add(rootViewDispatcher, queueInfo);
+
+                if (_active)
+                {
+                    // Simulate an OnResume from the correct dispatcher thread
+                    // (OnResume/OnSuspend/OnDestroy have this thread affinity, all other methods do enqueuings in a thread safe manner)
+                    DispatcherHelpers.RunOnDispatcher(rootViewDispatcher, queueInfo.queueInstance.OnResume, true); // inlining allowed
+                }
+            }
+            else
+            {
+                // Queue instance does exist.
+                // Increment the count of root views. This is helpful for the case the count reaches 0 so we can cleanup the queue.
+                queueInfo.rootViewCount++;
+            }
+
+            // Add tag
+            _reactTagToOperationQueue.Add(tag, queueInfo.queueInstance);
+
+            // Send forward
+            queueInfo.queueInstance.AddRootView(tag, rootView, themedRootContext);
+       }
+
+        /// <summary>
+        /// Enqueues an operation to remove the root view.
+        /// </summary>
+        /// <param name="rootViewTag">The root view tag.</param>
+        public void EnqueueRemoveRootView(int rootViewTag)
+        {
+            // Called on layout manager thread
+
+            UIViewOperationQueueInstance queue = GetQueueByTag(rootViewTag);
+
+            // Send forward
+            queue.EnqueueRemoveRootView(rootViewTag);
+
+            _reactTagToOperationQueue.Remove(rootViewTag);
+
+            // Do some maintenance/cleanup if needed.
+            // Find the queue info
+            var pair = _dispatcherToOperationQueueInfo.First(p => p.Value.queueInstance == queue);
+
+            // Decrement number of root views
+            pair.Value.rootViewCount--;
+
+            if (queue != MainUIViewOperationQueue)
+            {
+                if (pair.Value.rootViewCount == 0)
+                {
+                    // We can remove this queue and then destroy
+                    _dispatcherToOperationQueueInfo.Remove(pair.Key);
+
+                    // Simulate an OnDestroy from the correct dispatcher thread
+                    // (OnResume/OnSuspend/OnDestroy have this thread affinity, all other methods do enqueuings in a thread safe manner)
+                    DispatcherHelpers.RunOnDispatcher(pair.Key, queue.OnDestroy, true); // inlining allowed
+
+                    // Clean all the tags associated to this queue
+                    _reactTagToOperationQueue = _reactTagToOperationQueue.Where(p => p.Value != queue).ToDictionary(p => p.Key, p => p.Value);
```

Comment:
>_reactTagToOperationQueue [](start = 20, length = 25)

Is there any reason to create a new dictionary rather than just cleaning the entries out of the existing (and `readonly`) one?

Reply:
I just thought it would be faster, I'm not very experienced with cleaning a dictionary in place as you iterate through it. I can try.

---
In reply to: [166926394](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r166926394) [](ancestors = 166926394)

Reply:
The HashSet solution, with then a simple for each in the set to find the keys to remove from dict (no fancy functionals) is EXTREMELY fast compared with my original solution... Switching!

---
In reply to: [166984425](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r166984425) [](ancestors = 166984425,166926394)

Reply:
So, I'm torn. I would've gone with a Dictionary<queue, HashSet<tags>> approach to make the current behavior faster.
Problem is, I have some doubts with the current behavior. I looked closer and I realized JS side relies on the recursive cleanup of views.. You remove some views with "ManageChildren", all of the children and grandchildren are removed.
To keep the tag map clean and not growing I'd have to track this parent child tag relationship inside the queue collection... Basically the current "tag to queue" dictionary should become a "tag to <queue, list of children tags>" structure.

---
In reply to: [167003059](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167003059) [](ancestors = 167003059,166984425,166926394)

Reply:
Is there some way we can just hook into the recursion, so we don't have to build a complicated data structure? The parent/child relationships are already being stored *somewhere*, so it would be best if we could just hook into that.

---
In reply to: [167015776](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167015776) [](ancestors = 167015776,167003059,166984425,166926394)

Reply:
Would be easy to add `Action<int> onDropView` as a ctor parameter to NativeViewHierarchyManager and pass in an instance method on UIViewOperationQueue that cleans up whenever a view is dropped.

---
In reply to: [167195443](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167195443) [](ancestors = 167195443,167015776,167003059,166984425,166926394)

Reply:
You're very right. I was trying to avoid some elusive ghosts and I built a monster of my own. Not my very best day/decision.
I'll simplify, perhaps not today cause some other items caught up with me...

---
In reply to: [167197653](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167197653) [](ancestors = 167197653,167195443,167015776,167003059,166984425,166926394)